sys/net/application_layer/nanocoap/sock.c

```diff
@@ -131,6 +132,111 @@ ssize_t nanocoap_get(sock_udp_t *sock, const char *path, uint8_t *buf, size_t le
     return res;
 }
 
+static int _fetch_block(coap_pkt_t *pkt, uint8_t *buf, sock_udp_t *sock,
+                        const char *path, coap_blksize_t blksize, size_t num)
+{
+    uint8_t *pktpos = buf;
+    uint16_t lastonum = 0;
+
+    pkt->hdr = (coap_hdr_t *)buf;
+
+    pktpos += coap_build_hdr(pkt->hdr, COAP_TYPE_CON, NULL, 0, COAP_METHOD_GET,
+                             num);
+    pktpos += coap_opt_put_uri_pathquery(pktpos, &lastonum, path);
+    pktpos += coap_opt_put_uint(pktpos, lastonum, COAP_OPT_BLOCK2,
+                                (num << 4) | blksize);
+
+    pkt->payload = pktpos;
+    pkt->payload_len = 0;
+
+    int res = nanocoap_request(sock, pkt, NANOCOAP_BLOCKWISE_BUF(blksize));
+    if (res < 0) {
+        return res;
+    }
+
+    res = coap_get_code(pkt);
+    DEBUG("code=%i\n", res);
+    if (res != 205) {
+        return -res;
+    }
+
+    return 0;
+}
+
+int nanocoap_get_blockwise(sock_udp_t *sock, const char *path,
+                           coap_blksize_t blksize, void *buf,
+                           coap_blockwise_cb_t callback, void *arg)
+{
+    coap_pkt_t pkt;
+
+    int res, more = 1;
+    size_t num = 0;
+    res = -1;
+    while (more == 1) {
+        DEBUG("fetching block %u\n", (unsigned)num);
+        res = _fetch_block(&pkt, buf, sock, path, blksize, num);
+        DEBUG("res=%i\n", res);
+
+        if (!res) {
+            coap_block1_t block2;
+            coap_get_block2(&pkt, &block2);
+            more = block2.more;
+
+            if (callback(arg, block2.offset, pkt.payload, pkt.payload_len,
+                         more)) {
+                DEBUG("callback res != 0, aborting.\n");
+                res = -1;
+                goto out;
+            }
+        }
+        else {
+            DEBUG("error fetching block\n");
+            res = -1;
+            goto out;
+        }
+
+        num += 1;
+    }
+out:
+    return res;
+}
+
+int nanocoap_get_blockwise_url(const char *url,
+                               coap_blksize_t blksize, void *buf,
+                               coap_blockwise_cb_t callback, void *arg)
+{
+    char hostport[CONFIG_SOCK_HOSTPORT_MAXLEN];
+    char urlpath[CONFIG_SOCK_URLPATH_MAXLEN];
+    sock_udp_ep_t remote;
+    sock_udp_t sock;
+    int res;
+
+    if (strncmp(url, "coap://", 7)) {
```

Comment:
I vaguely remember we have a URI parsing library somewhere around.

Reply:
uh so you also want option processing here?

Reply:
No, I'm just saying that the splitting of a string into scheme, IP and path might be more reliably done in using [`uri_parser_process{,_string}`](https://doc.riot-os.org/group__sys__uri__parser.html#ga75b5c66965f3451eefb274569ce1147b), which not only not needs the external buffers, but also makes sure not to mix the path with the query, as this here appears to do.

Tracking this further, I do now notice that what says path here should actually be "path_and_query", so the "bug" that originally put me in this direction is not so much a bug as an oversimplification in names; that being so, `uri_parser_process` will not make our lives easier here as it splits up things that are later needed in a single bag. Please consider adjusting the names to match their use, and the suggestion in [here](https://github.com/RIOT-OS/RIOT/pull/17474/files/fcc46c596c0e5e81ce98d8fe7f606c199b2fc357..e8f1db52f48ca75053aee920d41744208e39c429#r791962596), and I'm happy with this part.